File: 0719-find-k-th-smallest-pair-distance/0719-find-k-th-smallest-pair-distance.py
class Solution:
    def smallestDistancePair(self, nums: List[int], k: int) -> int:
        n = len(nums)
        nums.sort()

        # max distance b/w two nums is dist between largest & smallest
        low, high = 0, nums[n-1] - nums[0]
        
        # binary search to find the distance at kth index
        while low < high:
            mid = low + (high - low) // 2
            count = self.countPairsWithMaxDistance(nums, mid)

            if count < k: low = mid+1
            else: high = mid

        return low
    
    def countPairsWithMaxDistance(self, nums: List[int], k: int) -> int:
        n, count = len(nums), 0

        left = 0
        for right in range(n):
            # increment left till distance > k
            while nums[right] - nums[left] > k: left += 1
            count += (right - left)

        return count

File: 0719-find-k-th-smallest-pair-distance/0719-find-k-th-smallest-pair-distance.cs
public class Solution {
    public int SmallestDistancePair(int[] nums, int k) {
        var n = nums.Length;
        Array.Sort(nums);

        // max distance b/w two nums is dist between largest & smallest
        int low = 0, high = nums[n-1] - nums[0];
        
        // binary search to find the distance at kth index
        while(low < high){
            var mid = low + (high - low) / 2;
            var count = CountPairsWithMaxDistance(nums, mid);

            if(count < k) low = mid+1;
            else high = mid;
        }

        return low;
    }

    private int CountPairsWithMaxDistance(int[] nums, int k){
        int n = nums.Length, count = 0;

        for(int left=0, right=0; right < n; right++){
            // increment left till distance > k
            while(nums[right] - nums[left] > k) left++;
            count += (right - left);
        }

        return count;
    }
}